libr/core/cmd_print.c

```diff
@@ -2658,27 +2719,70 @@ static int cmd_print(void *data, const char *input) {
 				ut32 bsz = core->blocksize;
 				RAnalFunction *f = r_anal_get_fcn_in (core->anal, core->offset,
 						R_ANAL_FCN_TYPE_FCN | R_ANAL_FCN_TYPE_SYM);
+				RListIter *iter;
+				RAnalBlock *bb;
+				RAnalFunction *tmp_func;
+				ut32 cont_size = 0;
+				RListIter *locs_it = NULL;
+				if (f && f->fcn_locs) {
+					locs_it = f->fcn_locs->head;
+					cont_size = r_anal_fcn_contsize (f);
+				}
 				if (f && input[2] == 'j') { // "pdfj"
-					ut8 *buf;
-					ut32 fcn_size = r_anal_fcn_size (f);
+					ut8 *func_buf = NULL, *loc_buf = NULL;
+					ut32 fcn_size = r_anal_fcn_realsize (f);
 					r_cons_printf ("{");
 					r_cons_printf ("\"name\":\"%s\"", f->name);
 					r_cons_printf (",\"size\":%d", fcn_size);
 					r_cons_printf (",\"addr\":%"PFMT64d, f->addr);
 					r_cons_printf (",\"ops\":");
 					// instructions are all outputted as a json list
-					buf = malloc (fcn_size);
-					if (buf) {
-						r_io_read_at (core->io, f->addr, buf, fcn_size);
-						r_core_print_disasm_json (core, f->addr, buf, fcn_size, 0);
-						free (buf);
+					func_buf = malloc (cont_size);
+					if (func_buf) {
+						//TODO: can loc jump to another locs?
+						for (locs_it; locs_it && (tmp_func = locs_it->data); locs_it = locs_it->n) {
+							if (tmp_func->addr < f->addr) {
+								cont_size = r_anal_fcn_contsize (tmp_func);
+								loc_buf = malloc (cont_size);
```

Comment:
use calloc to enforce some initialization, or check the r_io_read_at return value